File: board-room/src/main/java/net/stzups/board/room/server/websocket/protocol/server/messages/ServerMessageUpdateCanvas.java
package net.stzups.board.room.server.websocket.protocol.server.messages;

import io.netty.buffer.ByteBuf;
import net.stzups.board.data.objects.canvas.object.CanvasObjectType;
import net.stzups.board.data.objects.canvas.object.CanvasObjectWrapper;
import net.stzups.board.room.server.websocket.protocol.server.ServerMessage;
import net.stzups.board.room.server.websocket.protocol.server.ServerMessageType;

import java.util.Map;

public class ServerMessageUpdateCanvas extends ServerMessage {
    private Map<CanvasObjectType, Map<Short, CanvasObjectWrapper>> canvasObjects;

    public ServerMessageUpdateCanvas(Map<CanvasObjectType, Map<Short, CanvasObjectWrapper>> canvasObjects) {
        super(ServerMessageType.UPDATE_CANVAS);
        this.canvasObjects = canvasObjects;
    }

    public void serialize(ByteBuf byteBuf) {
        super.serialize(byteBuf);
        byteBuf.writeByte((byte) canvasObjects.size());
        for (Map.Entry<CanvasObjectType, Map<Short, CanvasObjectWrapper>> entry : canvasObjects.entrySet()) {
            byteBuf.writeByte((byte) entry.getKey().getId());
            byteBuf.writeShort((short) entry.getValue().size());
            for (Map.Entry<Short, CanvasObjectWrapper> entry1 : entry.getValue().entrySet()) {
                byteBuf.writeShort(entry1.getKey());
                entry1.getValue().serialize(byteBuf);
            }
        }
    }
}

File: scribbleshare-frontend/src/scripts/protocol/server/messages/ServerMessageUpdateCanvas.js
import ServerMessage from "../ServerMessage.js";
import ServerMessageType from "../ServerMessageType.js";
import CanvasObjectWrapper from "../../../canvas/CanvasObjectWrapper.js";

export default class ServerMessageUpdateCanvas extends ServerMessage {
    constructor(reader) {
        super(ServerMessageType.UPDATE_CANVAS);
        this.canvasObjectWrappers = new Map();
        let length = reader.readUint8();
        for (let i = 0; i < length; i++) {
            let type = reader.readUint8();
            let map = new Map();
            this.canvasObjectWrappers.set(type, map);
            let lengthJ = reader.readUint16();
            for (let j = 0; j < lengthJ; j++) {
                let id = reader.readInt16();
                let canvasObjectWrappers = [];
                let lengthK = reader.readUint8();
                for (let k = 0; k < lengthK; k++) {
                    canvasObjectWrappers[k] = new CanvasObjectWrapper(type, reader);
                }
                map.set(id, canvasObjectWrappers);
            }
        }
    }
}